core/src/verify.rs

```diff
@@ -1572,58 +1571,148 @@ mod test {
+            head: "Test reserved-diff commit".to_string(),
+            body: String::new(),
+            diff: Diff::Reserved(Box::new(reserved_state.clone())),
+        }))
+        .unwrap_err();
     }
 
     #[test]
-    fn test_verify_reserved_state_non_expelled_member() {
-        // configuring the test
-        let (mut validator_keypair, mut reserved_state, csv) = setup_test(4);
-        // adding a new member that will NOT be expelled
-        validator_keypair.push(generate_keypair([5]));
-        let new_expelled_member = Member {
-            public_key: validator_keypair.last().unwrap().0.clone(),
-            name: "new_non_expelled_member".to_string(),
+    /// Test the case where there is a duplicate public key.
+    fn invalid_reserved_state_with_duplicate_public_key() {
+        let (validator_keypair, mut reserved_state, mut csv) = setup_test(4);
+        // Generate a new member with duplicate public key
+        reserved_state.members.push(Member {
+            public_key: validator_keypair[0].0.clone(), // duplicate public key
+            name: format!("member{}", validator_keypair.len() - 1),
             governance_voting_power: 1,
             consensus_voting_power: 1,
             governance_delegatee: None,
             consensus_delegatee: None,
             expelled: false,
-        };
-        reserved_state.members.push(new_expelled_member.clone());
-        reserved_state
-            .consensus_leader_order
-            .push(new_expelled_member.name);
-
-        // rendering new reserved state that is expected to be valid with a non-expelled new member
-        let invalid_rs = reserved_state.clone();
+        });
+        // Apply reserved-diff commit to verify the reserved state
+        csv.apply_commit(&Commit::Transaction(Transaction {
+            author: "doesn't matter".to_owned(),
+            timestamp: 3,
+            head: "Test reserved-diff commit".to_string(),
+            body: String::new(),
+            diff: Diff::Reserved(Box::new(reserved_state.clone())),
+        }))
+        .unwrap_err();
+    }
 
-        assert!(csv.verify_reserved_state(&invalid_rs).is_ok());
+    #[test]
+    /// Test the case where the member names don't monotonically increase.
+    fn invalid_reserved_state_with_non_monotonic_increased_member_names() {
+        let (_, mut reserved_state, mut csv) = setup_test(5);
+        // Remove one from members instead of setting Member::expelled to true
+        reserved_state.members.pop();
+        reserved_state.consensus_leader_order.pop();
+        // Apply reserved-diff commit to verify the reserved state
+        csv.apply_commit(&Commit::Transaction(Transaction {
+            author: "doesn't matter".to_owned(),
+            timestamp: 3,
+            head: "Test reserved-diff commit".to_string(),
+            body: String::new(),
+            diff: Diff::Reserved(Box::new(reserved_state.clone())),
+        }))
+        .unwrap_err();
     }
```

Comment:
I think you might want to add the test case where any expelled members are included in the `consensus_leader_order` list.

```Rust
#[test]
    /// Test the case where an expelled member is included in the consensus leader order.
    fn invalid_reserved_state_with_expelled_member_in_consensus_leader_order() {
        let (_, mut reserved_state, mut csv) = setup_test(4);
        // Expel the first member
        reserved_state.members[0].expelled = true;
        // Include the expelled member in the consensus_leader_order
        reserved_state.consensus_leader_order.push(reserved_state.members[0].name.clone());
        // Apply reserved-diff commit to verify the reserved state
        csv.apply_commit(&Commit::Transaction(Transaction {
            author: "doesn't matter".to_owned(),
            timestamp: 3,
            head: "Test reserved-diff commit".to_string(),
            body: String::new(),
            diff: Diff::Reserved(Box::new(reserved_state)),
        }))
            .unwrap_err();
    }
```

Reply:
Thanks for the test suggestion! 
I added this test without the line `reserved_state.consensus_leader_order.push(reserved_state.members[0].name.clone());
 ` because `members[0]` is already included in `consensus_leader_order` by `setup_test()`